libc/picolibc: Don't enable malloc by default

keith-packard opened the pull request:
The common malloc implementation uses a heap which is initialized at system startup time. That gets linked into the application even if there are no references to malloc. This wastes memory and slows application startup.

When used with the minimal C library, this is "fixed" in the default configuration by setting the common malloc arena size to zero, removing the arena initialization code and providing stubs for malloc and free which are no-ops.

Picolibc adopted a different plan where a default malloc arena was provided when malloc was enabled. This ensures that malloc will be usable when it is available, which avoids a user trap where the application links but fails as soon as it tries to allocate any memory. However, that means that any application using picolibc will include the malloc arena initialization even if it doesn't use malloc. That costs a few hundred bytes of memory and a bit of startup time.

As an alternative, this patch series disables malloc entirely in the default configuration; appplications needing it will need to explicitly enable it in their configuration. All of the samples and tests which use malloc have their configurations updated.

Subsystems, drivers, libraries and modules which use malloc internally now signal their need for an allocator where necessary by selecting the `REQUIRES_MALLOC` Kconfig parameter.

keith-packard committed:
libc/picolibc: Only enable malloc when required

When malloc is included in the build, the heap initialization code will be
pulled in even if the application doesn't actually use malloc. This
increases the size of executables by a few hundred bytes and delays system
startup.

Select COMMON_LIBC_MALLOC only when applications indicate that they require
malloc by setting REQUIRES_MALLOC.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/minimal: Select common malloc when applications require it

Instead of relying on the default setting of `MINIMAL_LIBC_MALLOC` to pull
in the common malloc implementation, enable it whenever the application
indicates that it needs malloc through the REQUIRES_MALLOC setting.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
I've hacked this up to add a new config entry, `REQUIRES_MALLOC` and making that drive whether `COMMON_LIBC_MALLOC` is selected by both picolibc and the minimal C library. This will allow libraries, drivers, modules and other components to indicate a dependence on malloc without reference to the common malloc implementation -- directly enabling `COMMON_LIBC_MALLOC` will pull in that code even if it isn't the right malloc implementation.

keith-packard committed:
libc: Add REQUIRES_MALLOC to indirectly select malloc support

Instead of making applications use C library specific settings to enable
malloc support, provide this indirect symbol which then detects which C
library is in use and selects the correct configuration for each.

Signed-off-by: Keith Packard <keithp@keithp.com>

kartben commented:
> This will allow libraries, drivers, modules and other components to indicate a dependence on malloc 

And I think this PR should add such dependency to the likes of Thrift, TfLite, etc rather than doing it at the sample level, no?

keith-packard commented:
> > This will allow libraries, drivers, modules and other components to indicate a dependence on malloc
> 
> And I think this PR should add such dependency to the likes of Thrift, TfLite, etc rather than doing it at the sample level, no?

It probably won't be perfect, but yeah, it would be better to at least get started with this.

nordicjm commented on samples/modules/thrift/hello/client/prj.conf:
```diff
@@ -68,3 +68,5 @@ CONFIG_NET_TCP_WORKQ_STACK_SIZE=4096
 CONFIG_IDLE_STACK_SIZE=4096
 CONFIG_NET_RX_STACK_SIZE=8192
+
+CONFIG_REQUIRES_MALLOC=y
```
Yeah samples should not do this, @cfriedt would have to comment on adding it to the module Kconfig

nordicjm commented on samples/subsys/mgmt/mcumgr/smp_svr/overlay-fs.conf:
```diff
@@ -8,3 +8,5 @@ CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2304
 # Enable file system commands
 CONFIG_MCUMGR_GRP_FS=y
+
+CONFIG_REQUIRES_MALLOC=y
```
Likewise with this, should go to the subsys Kconfig, not the same, this would be a breaking change to every out of tree user using any of these modules/drivers/subsystems because suddenly after updating their applications would entirely just break if it is left like this

keith-packard committed:
libc: Add REQUIRED_MALLOC_ARENA_SIZE to indirectly request heap size

Instead of making applications use C library specific settings to specify
the desired minimum malloc arena size, provide this indirect symbol which
then detects which C library is in use and selects the correct
configuration for each.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
modules: Add REQUIRES_MALLOC to various modules

These modules use malloc APIs internally. Add REQUIRES_MALLOC to their
Kconfig settings to ensure the C library provides an implementation.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
net/sockets: Add REQUIRES_MALLOC when needed in sockets

The getaddrinfo function uses calloc when either a DNS resolver or an IP
stack is available. Indicate that by selecting REQUIRES_MALLOC when
necessary.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
lib/cpp: Imply REQUIRES_MALLOC when using C++

Most C++ applications will end up using various language features which
involve allocating memory, so most applications will need
malloc. Applications not needing it may still set REQUIRES_MALLOC=n if
they don't.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
fs/littlefs: Add REQUIRES_MALLOC when using littlefs

This module uses various malloc APIs internally.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/mem_alloc: Make tests labeled 'minimal' use the minimal C library

Now that picolibc is the default C library, tests which want to check
the minimal library need to explictly select it.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
samples,tests: Add REQUIRES_MALLOC=y to hash_map tests

These hash_map tests all use malloc directly to prepare various
test elements.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
ok, that wasn't as hard as I feared -- I think I've identified subsystems, modules and libraries that depend up on malloc and have placed `select REQUIRES_MALLOC=y` into their respective Kconfig files. There are still a couple of tests and samples which directly use malloc while the subsystem under test does not. Those still get `CONFIG_REQUIRES_MALLOC=y` added to their prj.conf files.

I'll iterate on this patch until CI is happy; local testing doesn't catch everything.

keith-packard committed:
soc/renasas_smartbond: Add REQUIRES_MALLOC to the da1469x configuration

The da1469x_pd driver uses calloc when reading trimv information.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
samples/ivshmem: Add CONFIG_REQUIRES_MALLOC=y

This sample uses malloc for a local buffer.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
This can definitely wait until after 3.5 is released.

jukkar reviewed (dismissed):
+1 to networking changes.

carlescufi commented:
@keith-packard please rebase

nordicjm reviewed (commented):
Not that I've really used it but wouldn't picolibc selecting this then prevent usage of custom malloc code (I know for zephyr that sounds incredibly unlikely) but for linux there are many different open source replacement malloc libraries

keith-packard commented:
> Not that I've really used it but wouldn't picolibc selecting this then prevent usage of custom malloc code (I know for zephyr that sounds incredibly unlikely) but for linux there are many different open source replacement malloc libraries

The picolibc configuration currently uses 'imply' here, not 'select', so applications are always free to disable use of the Zephyr common malloc implementation and use their own. All this series does is disable this default value so that applications don't get malloc unless necessary.

cfriedt commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -6,6 +6,7 @@
 menuconfig NET_SOCKETS
 	bool "BSD Sockets compatible API"
 	select FDTABLE
+	select REQUIRES_MALLOC if DNS_RESOLVER || NET_IP
```
technically, getaddrinfo and freeaddrinfo could maybe use a slab if malloc is unavailable. What do you think @rlubos ? Not a change required here, but maybe a good future enhancement. Even malloc-less systems should be able to resolve.

cfriedt commented on modules/thrift/Kconfig:
```diff
@@ -7,6 +7,7 @@ config ZEPHYR_THRIFT_MODULE
 menuconfig THRIFT
 	bool "Support for Thrift [EXPERIMENTAL]"
 	select EXPERIMENTAL
+	select REQUIRES_MALLOC
```
One day, It would be cool to not require a global allocator for Thrift

keith-packard commented:
Seems like this might uncover more malloc dependencies. Having malloc exist but always fail (with a zero-sized arena) seems like a user trap to me?

keith-packard committed:
modules/mcux: Add REQUIRES_MALLOC

This module uses malloc in the common code.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
samples/maxim_ds3231: Add REQUIRES_MALLOC (for strftime)

strftime deals with timezones and the picolibc (and newlib) implementations
of that code currently uses malloc. That will get fixed upstream
eventually, but for now, we need malloc to use it.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/cmsis_nn: Add REQUIRES_MALLOC

This test uses malloc for temporary buffers.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/mem_alloc: Skip minimal C library tests on full libc targets

Make sure these tests don't run when the underlying target requires a full
C library, and hence won't end up using the minimal C library anyways.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/basicmath: Add REQUIRES_MALLOC

This test uses malloc for output buffers.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/cmsis: Add REQUIRES_MALLOC

These tests allocate output buffers at runtime.

Signed-off-by: Keith Packard <keithp@keithp.com>

cfriedt reviewed (approved):
Still LGTM

keith-packard commented:
I've just posted PR #65434 which is an alternative to this PR that avoids including the malloc initialization code by not using `--whole-archive` on the common libc library. Would that be a better choice to reduce memory usage by applications not using malloc? Or would it be better to have applications explicitly request malloc when they need it?

keith-packard commented:
Ok, I think #65434 is the right way forward here -- applications do not have to declare whether they use malloc or not. When the don't, the malloc initialization code will not be included in the application.
